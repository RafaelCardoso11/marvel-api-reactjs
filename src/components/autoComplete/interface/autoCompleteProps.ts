export interface IAutoCompleteProps {
  placeholder?: string;
  value: string;
  setValueSearch: React.Dispatch<React.SetStateAction<string>>;
  handleAfterBounced: (param : any) => void;
  testID?: string
}
