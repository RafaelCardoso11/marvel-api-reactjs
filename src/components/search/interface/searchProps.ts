export interface ISearchProps {
    setValueAfterWrite: React.Dispatch<React.SetStateAction<string>> 
   }
   