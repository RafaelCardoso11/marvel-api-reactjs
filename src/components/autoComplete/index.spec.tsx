import { fireEvent, render } from "@testing-library/react";
import { AutoComplete } from "./index";
import { act } from "react-dom/test-utils";
import { IAutoCompleteProps } from "./interface/autoCompleteProps";

const AutoCompleteProps: IAutoCompleteProps = {
  handleAfterBounced: jest.fn(),
  setValueSearch: jest.fn(),
  value: "",
};

describe("<AutoComplete/>", () => {
  jest.useFakeTimers(); // Usar fake timers antes dos testes
  afterEach(() => {
    jest.clearAllTimers(); // Limpar todos os timers
  });

  it("Deveria renderizar na tela o componente Search com um Input", () => {
    const screen = render(<AutoComplete {...AutoCompleteProps} />);

    const searchInput = screen.getByTestId("id-search-characters");
    expect(searchInput).toBeInTheDocument();

    expect(screen.container).toMatchSnapshot();
  });
  it("Deveria ser possível buscar", () => {
    const handleAfterBounced = jest.fn();
    const valueSearch = "test";

    const setValue = jest.fn((valueRef) => {
      screen.rerender(
        <AutoComplete
          {...AutoCompleteProps}
          setValueSearch={setValue}
          handleAfterBounced={handleAfterBounced}
          value={valueRef}
        />
      );
    });
    const screen = render(
      <AutoComplete
        {...AutoCompleteProps}
        setValueSearch={setValue}
        handleAfterBounced={handleAfterBounced}
        value={""}
      />
    );

    const searchInput = screen.getByTestId("id-search-characters");
    fireEvent.change(searchInput, {
      target: { value: valueSearch },
    });

    act(() => {
      jest.runAllTimers();
    });

    expect(handleAfterBounced).toHaveBeenCalledTimes(1);
    expect(setValue).toHaveBeenCalled();

    expect(searchInput).toHaveValue(valueSearch);
    expect(screen.container).toMatchSnapshot();
  });
  it("Deveria ser possível buscar de forma Debounce", () => {
    const handleAfterBounced = jest.fn();
    const valueSearch = "teste";


    const setValue = jest.fn((valueRef) => {
      screen.rerender(
        <AutoComplete
          {...AutoCompleteProps}
          setValueSearch={setValue}
          handleAfterBounced={handleAfterBounced}
          value={valueRef}
        />
      );
    });

    const screen = render(
      <AutoComplete
        {...AutoCompleteProps}
        setValueSearch={setValue}
        handleAfterBounced={handleAfterBounced}
        value={""}
      />
    );

    const searchInput = screen.getByTestId("id-search-characters");

    fireEvent.change(searchInput, {
      target: { value: valueSearch },
    });

    act(() => {
      jest.runAllTimers();
    });
    expect(searchInput).toHaveValue(valueSearch);
    expect(handleAfterBounced).toHaveBeenCalledWith(valueSearch);
    expect(screen.container).toMatchSnapshot();
  });
});
