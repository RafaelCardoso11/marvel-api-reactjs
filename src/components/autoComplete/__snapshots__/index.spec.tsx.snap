// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AutoComplete/> Deveria renderizar na tela o componente Search com um Input 1`] = `
<div>
  <div
    class="sc-kvAqsT eAYqLr"
  >
    <input
      class="sc-grJzXs wpRIh"
      data-testid="id-search-characters"
      placeholder="Search..."
      type="text"
      value=""
    />
  </div>
</div>
`;

exports[`<AutoComplete/> Deveria renderizar um loading ao buscar 1`] = `
<div>
  <div
    class="sc-kvAqsT eAYqLr"
  >
    <input
      class="sc-grJzXs wpRIh"
      data-testid="id-search-characters"
      placeholder="Search..."
      type="text"
      value="test"
    />
  </div>
</div>
`;

exports[`<AutoComplete/> Deveria ser possível buscar 1`] = `
<div>
  <div
    class="sc-kvAqsT eAYqLr"
  >
    <input
      class="sc-grJzXs wpRIh"
      data-testid="id-search-characters"
      placeholder="Search..."
      type="text"
      value="test"
    />
  </div>
</div>
`;

exports[`<AutoComplete/> Deveria ser possível buscar de forma Debounce 1`] = `
<div>
  <div
    class="sc-kvAqsT eAYqLr"
  >
    <input
      class="sc-grJzXs wpRIh"
      data-testid="id-search-characters"
      placeholder="Search..."
      type="text"
      value="teste"
    />
  </div>
</div>
`;
